symengine/basic_to_upoly.cpp

```diff
@@ -0,0 +1,351 @@
+#include <symengine/visitor.h>
+
+namespace SymEngine
+{
+
+// returns the integer if `x` is a positive integer else
+// returns `-1` if negative integer, `0` if not an integer
+inline int positive_integer(const RCP<const Basic> &x)
+{
+    if (is_a<const Integer>(*x)) {
+        int tmp = rcp_static_cast<const Integer>(x)->as_int();
+        if (tmp > 0)
+            return tmp;
+        else
+            return -1;
+    }
+    return 0;
+}
+
+class UIntPolyGeneratorVisitor : public BaseVisitor<UIntPolyGeneratorVisitor>
+{
+private:
+    RCP<const Basic> res;
+    bool inpow;
+
+public:
+    RCP<const Basic> apply(const Basic &b, bool inpow_)
+    {
+        inpow = inpow_;
+        b.accept(*this);
+        return res;
+    }
+
+    void bvisit(const Pow &x)
+    {
+        if (is_a<const Rational>(*x.get_base())
+            or is_a<const Integer>(*x.get_base())) {
+            res = pow(x.get_base(), _find_gen_uintpoly(x.get_exp(), true));
+        } else if (is_a<const Rational>(*x.get_exp())
+                   or is_a<const Integer>(*x.get_exp())) {
+
+            int tmp = positive_integer(x.get_exp());
+            if (tmp != 0) {
+                if (tmp > 0)
+                    res = _find_gen_uintpoly(x.get_base());
+                else
+                    res = div(one, x.get_base());
+                return;
+            }
+
+            if (is_a<const Rational>(*x.get_exp())) {
+                RCP<const Rational> rat
+                    = rcp_static_cast<const Rational>(x.get_exp());
+                if (rat->is_negative())
+                    res = pow(x.get_base(), div(neg(one), rat->get_den()));
+                else
+                    res = pow(x.get_base(), div(one, rat->get_den()));
+            } else {
+                throw std::runtime_error("Could not extract generator");
+            }
+        } else {
+            // won't handle cases like genarator = x**x
+            throw std::runtime_error("Could not extract generator");
+        }
+    }
+
+    void _update_min_gen(RCP<const Basic> &min_gen, RCP<const Basic> &curr_gen, bool &min_gen_set)
+    {
+        if (min_gen_set) {
+            if ((is_a_sub<const Function>(*min_gen)
+                 and is_a_sub<const Function>(*curr_gen))
```

Comment:
I think this logic can be made simpler by,
1. Check if `min_gen` and `curr_gen` are both not Pows. In this case they need to equal
2. Create `base1, exp1, base2, exp2` where `base1, exp1 = min_gen.base(), min_gen.exp()` if `min_gen` is a `Pow`, else `base1, exp1 = min_gen, 1`. Same for curr_gen. Handle this case all at once